walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,81 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    EventsTimeRange eventsTimeRange = filterEventsTime(events, request.getAttendees(), request.getOptionalAttendees());
+
+    Collection<TimeRange> times = findTimes(eventsTimeRange.getAllEventsTimeRange(), request.getDuration());
+    if (times.isEmpty()) {
+      System.out.println(eventsTimeRange.getMandatoryEventsTimeRange());
+      times = findTimes(eventsTimeRange.getMandatoryEventsTimeRange(), request.getDuration());
+    }
+    return times;
+  }
+  
+  private EventsTimeRange filterEventsTime(Collection<Event> events, Collection<String> attendees, Collection<String> optional_attendees) {
+    Collection<TimeRange> mandatoryEventsTimeRange = new HashSet<>();
+    Collection<TimeRange> optionalEventsTimeRange = new HashSet<>();
+
+    for (Event event : events) {
+      Collection<String> event_attendees = event.getAttendees();
+      if (!intersection(event_attendees, attendees).isEmpty()) {
+        mandatoryEventsTimeRange.add(event.getWhen());
+      }
+      if (!intersection(event_attendees, optional_attendees).isEmpty()) {
+        optionalEventsTimeRange.add(event.getWhen());
+      }
+    }
+    
+    return new EventsTimeRange(mandatoryEventsTimeRange, optionalEventsTimeRange);
+  }
+
+  private Collection<String> intersection(Collection<String> c1, Collection<String> c2) {
+    Collection<String> intersection = new HashSet<String>(c1);
+    intersection.retainAll(c2);
+    return intersection;
+  }
+
+  private Collection<TimeRange> findTimes(Collection<TimeRange> eventsTimeRange, long duration) {
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    if (eventsTimeRange.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    List<TimeRange> eventsTimeRangeList = new ArrayList<>(eventsTimeRange);
+    Collections.sort(eventsTimeRangeList, TimeRange.ORDER_BY_START);
+
+    Collection<TimeRange> times = new ArrayList<>();
+    int end = TimeRange.START_OF_DAY;
+    for (TimeRange eventTimeRange : eventsTimeRangeList) {
+      int start = eventTimeRange.start();
+      if (end < start) {
+        if (start - end >= duration) {
+          times.add(TimeRange.fromStartEnd(end, start, false));
```

Comment:
The boolean parameter here makes the code hard to understand, because I don't know what property it modifies. 
One way to remedy this is to introduce a comment like this:

```
times.add(TimeRange.fromStartEnd(end, start, /* inclusive= */ false));
```